sensors/2.0/iiohal_mediation_v2.0/libiio_client/network.c

```diff
@@ -1625,8 +1666,145 @@ struct iio_context * network_create_context(const char *host)
 err_free_pdata:
     free(pdata);
 err_close_socket:
-    close(fd);
+	close(fd);
+err_exit:
+	return NULL;
+}
+
+struct iio_context * network_create_context(const char *host)
+{
+	struct addrinfo hints, *res;
+	int ret;
+	struct iio_context *ctx;
+#ifdef _WIN32
+	WSADATA wsaData;
+
+	ret = WSAStartup(MAKEWORD(2, 0), &wsaData);
+	if (ret < 0) {
+		ERROR("WSAStartup failed with error %i\n", ret);
+		errno = -ret;
+		return NULL;
+	}
+#endif
+
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+
+#ifdef HAVE_AVAHI
+	if (!host) {
+		char addr_str[AVAHI_ADDRESS_STR_MAX];
+		char port_str[6];
+		AvahiAddress address;
+		uint16_t port = IIOD_PORT;
+
+		memset(&address, 0, sizeof(address));
+
+		ret = discover_host(&address, &port);
+		if (ret < 0) {
+			char buf[1024];
+			iio_strerror(-ret, buf, sizeof(buf));
+			DEBUG("Unable to find host: %s\n", buf);
+			errno = -ret;
+			return NULL;
+		}
+
+		avahi_address_snprint(addr_str, sizeof(addr_str), &address);
+		iio_snprintf(port_str, sizeof(port_str), "%hu", port);
+		ret = getaddrinfo(addr_str, port_str, &hints, &res);
+	} else
+#endif
+	{
+		ret = getaddrinfo(host, IIOD_PORT_STR, &hints, &res);
+	}
+
+	if (ret) {
+		ERROR("Unable to find host: %s\n", gai_strerror(ret));
+#ifndef _WIN32
+		if (ret != EAI_SYSTEM)
+			errno = -ret;
+#endif
+		return NULL;
+	}
+
+	ctx = do_network_create_context(res);
+	if (ctx == NULL)
+		goto err_free_addrinfo;
+	else
+		return ctx;
+
 err_free_addrinfo:
     freeaddrinfo(res);
     return NULL;
 }
+
+static struct iio_context * sock_create_context(sa_family_t family,
+						struct sockaddr* addr,
+						size_t addrlen)
+{
+	struct addrinfo *ai = zalloc(sizeof(*ai));
+
+	if (!ai) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	ai->ai_family = addr->sa_family = family;
+	ai->ai_socktype = SOCK_STREAM;
+
+	ai->ai_addr = addr;
+	ai->ai_addrlen = addrlen;
+
+	return do_network_create_context(ai);
+}
+
+struct iio_context * unix_create_context(const char *path)
+{
+	struct iio_context *ctx;
+	struct sockaddr_un *addr = zalloc(sizeof(*addr));
+
+	if (!addr) {
+		errno = ENOMEM;
+		goto err_exit;
+	}
+
+	strncpy(addr->sun_path, path, sizeof(addr->sun_path)-1);
+
+	ctx = sock_create_context(AF_UNIX, (struct sockaddr*)addr,
+				  sizeof(*addr));
+	if (ctx == NULL)
+		goto err_free;
+	else
+		return ctx;
+
+err_free:
+	free(addr);
+err_exit:
+	return NULL;
+}
+
+struct iio_context * vm_create_context(unsigned int port)
+{
+	struct iio_context *ctx;
+	struct sockaddr_vm *addr = zalloc(sizeof(*addr));
```

Comment:
Once the socket is created, do we still require this memory?
If not pls free the memory or else we could directly declare non-pointer variable.